tools/notification-configuration/notification-creator/Contacts.cs

```diff
@@ -0,0 +1,126 @@
+/// </summary>
+internal class Contacts
+{
+    private readonly ILogger log;
+    private readonly GitHubService gitHubService;
+
+    // Type 2 maps to a build definition YAML file in the repository
+    private const int BuildDefinitionYamlProcessType = 2;
+
+    internal Contacts(GitHubService gitHubService, ILogger log)
+    {
+        this.log = log;
+        this.gitHubService = gitHubService;
+    }
+
+    /// <summary>
+    /// See the class comment.
+    /// </summary>
+    public async Task<List<string>> GetFromBuildDefinitionRepoCodeowners(BuildDefinition buildDefinition)
+    {
+        if (buildDefinition.Process.Type != BuildDefinitionYamlProcessType)
+        {
+            this.log.LogDebug(
+                "buildDefinition.Process.Type for buildDefinition.Name = '{buildDefinitionName}' " +
+                "is not BuildDefinitionYamlProcessType = {BuildDefinitionYamlProcessType}.",
+                buildDefinition.Name,
+                BuildDefinitionYamlProcessType);
+            return null;
+        }
+
+        Uri repoUrl = GetCodeownersRepoUrl(buildDefinition);
+
+        if (repoUrl == null)
+        {
+            // assert: the reason why repoUrl is null has been already logged.
+            return null;
+        }
+
+        List<CodeownersEntry> codeownersEntries = await gitHubService.GetCodeownersFileEntries(repoUrl);
+
+        if (codeownersEntries == default)
+        {
+            this.log.LogInformation("CODEOWNERS file in '{repoUrl}' not found, skipping sync.", repoUrl);
+            return null;
+        }
+
+        if (buildDefinition.Process is not YamlProcess process)
+        {
+            this.log.LogError(
+                "buildDefinition.Process as YamlProcess is null. buildDefinition.Name: '{buildDefinitionName}'",
+                buildDefinition.Name);
+            return null;
+        }
+
+        // process.YamlFilename is misleading here. It is actually a file path, not file name.
+        // E.g. it is "sdk/foo_service/ci.yml".
+        string buildDefinitionFilePath = process.YamlFilename; 
+
+        this.log.LogInformation(
+            "Searching CODEOWNERS for matching path for '{buildDefinitionFilePath}'",
+            buildDefinitionFilePath);
+
+        CodeownersEntry matchingCodeownersEntry = GetMatchingCodeownersEntry(process, codeownersEntries);
+        List<string> contacts = matchingCodeownersEntry.Owners;
+
+        this.log.LogInformation(
+            "Found matching contacts (owners) in CODEOWNERS. " +
+            "Searched path = '{buildDefinitionOwnerFile}', Contacts# = {contactsCount}",
+            buildDefinitionFilePath,
+            contacts.Count);
+
+        return contacts;
+    }
+
+    private Uri GetCodeownersRepoUrl(BuildDefinition buildDefinition)
+    {
+        Uri repoUrl = buildDefinition.Repository.Url;
+        this.log.LogInformation("Fetching CODEOWNERS file from repoUrl: '{repoUrl}'", repoUrl);
+
+        if (repoUrl != null)
```

Comment:
Not sure what the schema says, but this seems a little more defensive in the case of empty string:

```suggestion
        if (String.IsNullOrEmpty(repoUrl))
```